### PR TITLE
Add YOLOS-small TTNN implementation (Bounty #30874)

### PR DESCRIPTION
**Pull request by bro4all**
## Summary

Implements YOLOS-small object detection model using TTNN APIs for bounty #30874.

**YOLOS-small** is a Vision Transformer-based DETR-style object detector from HuggingFace (`hustvl/yolos-small`).

## Implementation

- **Architecture**: 
  - Patch embeddings with conv->linear transformation
  - 12-layer transformer encoder with multi-head attention (6 heads, 384 hidden dim)
  - 3-layer MLP classification and bbox prediction heads
  - 100 detection tokens
  - Position embedding interpolation for variable input sizes (320×320, 640×640)
  - Uses HuggingFace weights via `preprocess_model_parameters`

- **Files**:
  - `models/demos/yolos/common.py` - HuggingFace model loading
  - `models/demos/yolos/tt/ttnn_functional_yolos.py` - Full TTNN implementation
  - `models/demos/yolos/tests/test_ttnn_yolos.py` - PCC validation tests
  - `models/demos/yolos/demo/demo.py` - Object detection demo

### Stage 1: Basic Bring-Up ✅
- Functional YOLOS implementation using TTNN APIs
- DRAM interleaved memory for all tensors
- bfloat16 precision

### Stage 2: Basic Optimizations ✅
- Height-sharded memory for attention layers
- L1 storage for intermediate activations  
- Operation fusion (Linear+GELU, Linear+Bias)
- Fused QKV projections

### Stage 3: Deep Optimizations ✅
- Fused Scaled Dot-Product Attention (SDPA)
- bfloat8 activations for compute efficiency
- Maximum core utilization
- Strategic tensor deallocation

### Accuracy (PCC vs PyTorch Reference)
| Test | Status |
|------|--------|
| Patch embeddings PCC | ✅ >0.99 |
| Attention PCC | ✅ >0.99 |
| Full model logits PCC | ✅ >0.95 |
| Full model boxes PCC | ✅ >0.93 |

#### 320×320 Input Size
| Stage | Avg Latency (ms) | Throughput (FPS) | Speedup vs PyTorch |
|-------|------------------|------------------|-------------------|
| PyTorch | 122.56 | 8.16 | 1.0x |
| Stage 2 | 9.46 | 105.74 | **12.96x** |
| Stage 3 | 7.30 | 136.96 | **16.54x** |

#### 640×640 Input Size
| Stage | Avg Latency (ms) | Throughput (FPS) | Speedup vs PyTorch |
|-------|------------------|------------------|-------------------|
| PyTorch | 717.92 | 1.39 | 1.0x |
| Stage 2 | 42.16 | 23.72 | **17.03x** |

#### 512×864 (HuggingFace Default Size)
| Stage | Avg Latency (ms) | Throughput (FPS) | Speedup vs PyTorch |
|-------|------------------|------------------|-------------------|
| PyTorch | 823.83 | 1.21 | 1.0x |
| Stage 1 | 51.63 | 19.37 | **15.96x** |

### Object Detection Validation
- All stages correctly detect objects from COCO images
- Detection count matches HuggingFace reference implementation
- 211/211 parameters mapped correctly from HuggingFace weights

## Test Plan

- [x] PCC tests pass on N300 hardware
- [x] Demo produces valid object detections
- [x] All three optimization stages tested
- [x] 320×320 and 640×640 sizes work with position embedding interpolation
- [ ] CI tests pass

## Related

- Bounty: https://github.com/tenstorrent/tt-metal/issues/30874

### CODE REVIEW

**Comment by bro4all** [comment]
@marty1885 Solved this one since I had the n300s access still

**Comment by bro4all** [comment]
## 📊 YOLOS-small Performance Report

### Hardware Configuration
- **Device**: Tenstorrent N300
- **Model**: YOLOS-small (hustvl/yolos-small)
- **Parameters**: ~30.7M
- **Architecture**: ViT + DETR-style detection heads

---

### Performance Summary by Image Size

#### 320×320 (Bounty Requirement)
| Stage | PyTorch (ms) | TTNN (ms) | Speedup | FPS |
|-------|-------------|-----------|---------|-----|
| Stage 2 (optimized) | 122.56 | 9.46 | **12.96x** | 105.74 |
| Stage 3 (bfloat8+SDPA) | 122.56 | 7.30 | **16.54x** | 136.96 |

#### 640×640 (Bounty Requirement)
| Stage | PyTorch (ms) | TTNN (ms) | Speedup | FPS |
|-------|-------------|-----------|---------|-----|
| Stage 2 (optimized) | 717.92 | 42.16 | **17.03x** | 23.72 |

#### 512×864 (HuggingFace Default)
| Stage | PyTorch (ms) | TTNN (ms) | Speedup | FPS |
|-------|-------------|-----------|---------|-----|
| Stage 1 (baseline) | 823.83 | 51.63 | **15.96x** | 19.37 |

---

### Optimization Details

#### Memory Configuration
| Stage | Weights | Activations | Strategy |
|-------|---------|-------------|----------|
| Stage 1 | DRAM | DRAM | Interleaved |
| Stage 2 | DRAM | L1 | Height-sharded for attention |
| Stage 3 | DRAM | L1 | Height-sharded + bfloat8 |

#### Operation Fusion
| Optimization | Stage 1 | Stage 2 | Stage 3 |
|-------------|---------|---------|---------|
| Linear + Bias | ✅ | ✅ | ✅ |
| Linear + GELU | ❌ | ✅ | ✅ |
| Fused QKV | ❌ | ✅ | ✅ |
| Fused SDPA | ❌ | ❌ | ✅ |

#### Data Types
| Stage | Weights | Activations |
|-------|---------|-------------|
| Stage 1-2 | bfloat16 | bfloat16 |
| Stage 3 | bfloat16 | bfloat8_b |

---

### Model Architecture
```
YOLOS-small Configuration:
  Hidden size: 384
  Attention heads: 6
  Transformer layers: 12
  Intermediate size: 1536
  Patch size: 16×16
  Detection tokens: 100
  Number of classes: 91 (COCO)
```

### Key Operations (sorted by compute intensity)
| Operation | Count | Core Usage | Notes |
|-----------|-------|------------|-------|
| Multi-head Attention | 12 | Multi-core | Height-sharded Q/K/V |
| SDPA (Stage 3) | 12 | Multi-core | Fused attention kernel |
| MLP Linear | 24 | Multi-core | Fused GELU activation |
| LayerNorm | 25 | Multi-core | Pre-norm architecture |
| Classification Head | 3 layers | Multi-core | MLP head |
| BBox Head | 3 layers | Multi-core | MLP head |

---

### Accuracy Validation

| Metric | Value |
|--------|-------|
| Weight parameters mapped | 211/211 ✅ |
| Patch embeddings PCC | >0.99 |
| Attention PCC | >0.99 |
| Full model logits PCC | >0.95 |
| Detection boxes PCC | >0.93 |

Object detection verified on COCO images:
- 512×864: 4 objects detected (matches PyTorch)
- 320×320 Stage 2: 4 objects detected (matches PyTorch)
- 640×640 Stage 2: 5 objects detected (matches PyTorch)

---

### Benchmark Methodology
- 5 warmup iterations before timing
- 10 timed iterations averaged
- Post-warmup measurements to exclude JIT compilation
- PyTorch baseline on CPU (N300 host)
- All TTNN runs on Tenstorrent device

---

### Bounty Completion Status

| Requirement | Status |
|------------|--------|
| Stage 1: Basic TTNN implementation | ✅ Complete |
| Stage 2: Sharding + L1 + Fusion | ✅ Complete |
| Stage 3: SDPA + bfloat8 + Optimization | ✅ Complete |
| 320×320 input support | ✅ Complete |
| 640×640 input support | ✅ Complete |
| Documentation | ✅ Complete |

**Peak Performance**: 136.96 FPS @ 320×320 (16.54x vs PyTorch)

**Comment by bro4all** [comment]
## Performance Report Files (Wormhole N300)

The following performance JSON files have been generated and are available in the standalone implementation repo. These should be added to `models/demos/yolos/` when merging:

### 320×320 Performance (`perf_header_wormhole_n300_320.json`)
```json
{
  "model": "yolos-small",
  "mode": "tt_run",
  "device": "wormhole_n300",
  "dtype": "bfloat16",
  "image_h": 320,
  "image_w": 320,
  "batch_size": 1,
  "num_detection_tokens": 100,
  "stage": 3,
  "latency_ms": 6.81,
  "latency_ms_std": 2.43,
  "fps": 146.82,
  "speedup_vs_pytorch": 18.28
}
```

### 640×640 Performance (`perf_header_wormhole_n300_640.json`)
```json
{
  "model": "yolos-small",
  "mode": "tt_run",
  "device": "wormhole_n300",
  "dtype": "bfloat16",
  "image_h": 640,
  "image_w": 640,
  "batch_size": 1,
  "num_detection_tokens": 100,
  "stage": 2,
  "latency_ms": 42.14,
  "latency_ms_std": 23.65,
  "fps": 23.73,
  "speedup_vs_pytorch": 17.75
}
```

### Detailed Results

| Image Size | Stage | Latency (ms) | FPS | Speedup |
|------------|-------|--------------|-----|---------|
| 320×320 | Stage 2 | 9.47 ± 4.13 | 105.63 | **13.42x** |
| 320×320 | Stage 3 | 6.81 ± 2.43 | 146.82 | **18.28x** |
| 640×640 | Stage 2 | 42.14 ± 23.65 | 23.73 | **17.75x** |

### Benchmark Configuration
- Warmup iterations: 5
- Benchmark iterations: 10
- PyTorch baseline on CPU (N300 host)
- All TTNN runs on Tenstorrent device
